ambry-network/src/test/java/com.github.ambry.network/ConnectionTrackerTest.java

```diff
@@ -205,6 +203,76 @@ public void testConnectionTracker() {
     assertCounts(totalConnectionsCount, availableCount);
   }
 
+  /**
+   * Test behavior of {@link ConnectionTracker#setMinimumActiveConnectionsPercentage} and
+   * {@link ConnectionTracker#replenishConnections}.
+   * @throws IOException
+   */
+  @Test
+  public void testReplenishConnections() {
+    connectionTracker = new ConnectionTracker(routerConfig.routerScalingUnitMaxConnectionsPerPortPlainText,
+        routerConfig.routerScalingUnitMaxConnectionsPerPortSsl);
+    // When no connections were ever made to a host:port, connectionTracker should return null, but
+    // initiate connections.
+    int minActiveConnectionsCount = 0;
+    int totalConnectionsCount = 0;
+    int availableCount = 0;
+
+    MockDataNodeId dataNodeId1 =
+        new MockDataNodeId("host1", Collections.singletonList(plainTextPort), Collections.emptyList(), "DC1");
+    MockDataNodeId dataNodeId2 =
+        new MockDataNodeId("host2", Arrays.asList(plainTextPort, sslPort), Collections.emptyList(), "DC1");
+    dataNodeId2.setSslEnabledDataCenters(Collections.singletonList("DC1"));
+    connectionTracker.setMinimumActiveConnectionsPercentage(dataNodeId1, 50);
+    minActiveConnectionsCount += 50 * routerConfig.routerScalingUnitMaxConnectionsPerPortPlainText / 100;
+    connectionTracker.setMinimumActiveConnectionsPercentage(dataNodeId2, 200);
+    minActiveConnectionsCount += routerConfig.routerScalingUnitMaxConnectionsPerPortSsl;
+
+    // call replenishConnections to warm up connections
+    assertCounts(totalConnectionsCount, availableCount);
+    connectionTracker.replenishConnections(this::mockNewConnection);
+    totalConnectionsCount += minActiveConnectionsCount;
+    assertCounts(totalConnectionsCount, availableCount);
+    List<String> newConnections = getNewlyEstablishedConnections();
+    newConnections.forEach(connectionTracker::checkInConnection);
+    availableCount += minActiveConnectionsCount;
+    assertCounts(totalConnectionsCount, availableCount);
+    Assert.assertTrue(connectionTracker.mayCreateNewConnection("host1", plainTextPort, dataNodeId1));
+    Assert.assertFalse(connectionTracker.mayCreateNewConnection("host2", sslPort, dataNodeId2));
+
+    // remove 2 connections
+    newConnections.stream().limit(2).forEach(connectionTracker::removeConnection);
+    totalConnectionsCount -= 2;
+    availableCount -= 2;
+    assertCounts(totalConnectionsCount, availableCount);
```

Comment:
it doesn't seem like this is really testing anything after line 239 (I don't see where line 244 affects `totalConnectionsCount` or `availableCount`)

Reply:
244 takes 2 connections and removes them from the connection pool, thus decreasing the connection count and connections available to check out by 2. This is mostly preparation for the replenishConnections call on line 250, but I wanted to ensure that the connection count was as expected before that call is made.

Reply:
oh I see, `assertCounts` looks at newConnections.  Sorry, I had misread it and thought it was just comparing the two integers